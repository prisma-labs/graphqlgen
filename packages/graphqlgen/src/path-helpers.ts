import * as fs from 'fs'
import * as path from 'path'
import { Language } from 'graphqlgen-json-schema'

export function getExtNameFromLanguage(language: Language) {
  const extNames = {
    typescript: '.ts',
    flow: '.js',
  }

  return extNames[language]
}

export function getAbsoluteFilePath(
  modelPath: string,
  language: Language,
): string {
  let absolutePath = path.resolve(modelPath)
  const extName = getExtNameFromLanguage(language)

  if (
    !fs.existsSync(absolutePath) &&
    fs.existsSync(`${absolutePath}${extName}`)
  ) {
    absolutePath += extName
  }

  if (!fs.existsSync(absolutePath)) {
    throw new Error(`${absolutePath} not found`)
  }

  if (!fs.lstatSync(absolutePath).isDirectory()) {
    if (path.extname(absolutePath) !== extName) {
      throw new Error(`${absolutePath} has to be a ${extName} file`)
    }

    return absolutePath.replace(/\\/g, '/')
  }

  const indexPath = path.join(absolutePath, 'index' + extName)
  if (!fs.existsSync(indexPath)) {
    throw new Error(
      `No index${extName} file found in directory: ${absolutePath}`,
    )
  }

  return indexPath.replace(/\\/g, '/')
}

export function getImportPathRelativeToOutput(
  absolutePath: string,
  outputDir: string,
): string {
  let relativePath = path.relative(path.dirname(outputDir), absolutePath)

  if (!relativePath.startsWith('.')) {
    relativePath = './' + relativePath
  }

  // remove .ts or .js file extension
  relativePath = relativePath.replace(/\.(ts|js)$/, '')

  // remove /index
  relativePath = relativePath.replace(/\/index$/, '')

  // replace \ with /
  relativePath = relativePath.replace(/\\/g, '/')

  return relativePath
}
