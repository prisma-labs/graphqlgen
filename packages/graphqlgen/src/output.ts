import chalk from 'chalk'
import * as os from 'os'
import { graphQLToTypecriptFlowType, GraphQLTypeObject } from './source-helper'
import { maybeReplaceDefaultName, ValidatedDefinition } from './validation'
import { Language } from 'graphqlgen-json-schema'
import { getExtNameFromLanguage } from './path-helpers'

export function outputDefinitionFilesNotFound(
  validatedDefinitions: ValidatedDefinition[],
) {
  const invalidDefinitions = validatedDefinitions.filter(
    validation => !validation.fileExists,
  )

  console.log(`❌ Some path to model definitions defined in your graphqlgen.yml were not found
  
${chalk.bold(
  'Step 1',
)}: Make sure each of these model definitions point to an existing file

  models:
    override:
    ${invalidDefinitions
      .map(
        def =>
          `  ${def.definition.typeName}: ${chalk.redBright(
            def.definition.filePath!,
          )}:${def.definition.modelName}`,
      )
      .join(os.EOL)}

${chalk.bold('Step 2')}: Re-run ${chalk.bold('`graphqlgen`')}`)
}

export function outputWrongSyntaxFiles(
  validatedDefinitions: ValidatedDefinition[],
) {
  const invalidDefinitions = validatedDefinitions.filter(
    validation => !validation.validSyntax,
  )

  console.log(`❌ Some model definitions defined in your graphqlgen.yml have syntax errors
  
${chalk.bold(
  'Step 1',
)}: Make sure each of these model definitions follow the correct syntax

  ${chalk.cyan(
    `(Correct syntax: ${chalk.bold('<typeName>')}: ${chalk.bold(
      '<filePath>',
    )}:${chalk.bold('<modelName>')})`,
  )}

  models:
    override:
    ${invalidDefinitions
      .map(def =>
        chalk.redBright(
          `  ${def.definition.typeName}: ${def.definition.rawDefinition}`,
        ),
      )
      .join(os.EOL)}

${chalk.bold('Step 2')}: Re-run ${chalk.bold('`graphqlgen`')}`)
}

export function outputInterfaceDefinitionsNotFound(
  validatedDefinitions: ValidatedDefinition[],
) {
  const invalidDefinitions = validatedDefinitions.filter(
    validation => !validation.interfaceExists,
  )

  console.log(`❌ Some model definitions defined in your graphqlgen.yml were not found
  
${chalk.bold(
  'Step 1',
)}: Make sure each of these model definitions are defined in the following files

  models:
    override:
    ${invalidDefinitions
      .map(
        def =>
          `  ${def.definition.typeName}: ${
            def.definition.filePath
          }:${chalk.redBright(def.definition.modelName!)}`,
      )
      .join(os.EOL)}

${chalk.bold('Step 2')}: Re-run ${chalk.bold('`graphqlgen`')}`)
}

export function outputMissingModels(
  missingModels: GraphQLTypeObject[],
  language: Language,
  defaultName: string | null,
) {
  console.log(`❌ Some types from your application schema have model definitions that are not defined yet
  
${chalk.bold(
  'Step 1',
)}: Copy/paste the model definitions below to your application

${missingModels
  .map(type => renderModelFromType(type, language, defaultName))
  .join(os.EOL)}


${chalk.bold('Step 2')}: Reference the model definitions in your ${chalk.bold(
    'graphqlgen.yml',
  )} file

models:
  files:
    - ./path/to/your/file${getExtNameFromLanguage(language)}  

${chalk.bold('Step 3')}: Re-run ${chalk.bold('`graphqlgen`')}`)
}

export function outputModelFilesNotFound(filesNotFound: string[]) {
  console.log(`❌ Some path to model definitions defined in your graphqlgen.yml were not found
  
${chalk.bold('Step 1')}: Make sure each of these files exist

  models:
    files:
    ${filesNotFound.map(file => `  - ${chalk.redBright(file)}`).join(os.EOL)}

${chalk.bold('Step 2')}: Re-run ${chalk.bold('`graphqlgen`')}`)
}

function renderModelFromType(
  type: GraphQLTypeObject,
  language: Language,
  defaultName: string | null,
): string {
  switch (language) {
    case 'typescript':
      return renderTypescriptModelFromType(type, defaultName)
    case 'flow':
      return renderFlowModelFromType(type, defaultName)
  }
}

function renderTypescriptModelFromType(
  type: GraphQLTypeObject,
  defaultName: string | null,
): string {
  const name = maybeReplaceDefaultName(type.name, defaultName)
  return `\
export interface ${chalk.bold(name)} {
${type.fields
  .filter(field => field.type.isScalar)
  .map(field => `  ${field.name}: ${graphQLToTypecriptFlowType(field.type)}`)
  .join(os.EOL)}
}`
}

function renderFlowModelFromType(
  type: GraphQLTypeObject,
  defaultName: string | null,
): string {
  const name = maybeReplaceDefaultName(type.name, defaultName)
  return `\
export interface ${chalk.bold(name)} {
${type.fields
  .map(field => `  ${field.name}: ${graphQLToTypecriptFlowType(field.type)}`)
  .join(',' + os.EOL)}
}`
}
