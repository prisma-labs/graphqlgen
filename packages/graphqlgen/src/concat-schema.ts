import chalk from 'chalk'
import * as fs from 'fs'
import { GraphQLGenDefinition } from 'graphqlgen-json-schema'
import * as mkdirp from 'mkdirp'
import * as path from 'path'

import { extractGlobPattern } from './glob'

export const readSchemaSync = (schemaDir: string) => {
  try {
    const schemas = extractGlobPattern([schemaDir])

    if (schemas) {
      return schemas
        .map(f => {
          return fs.readFileSync(f).toString()
        })
        .join('\n')
    }
  } catch (error) {
    console.log(error)
  }
}

export const generateSchema = (config: GraphQLGenDefinition) => {
  const data = readSchemaSync(config.schema)

  const outputPath = path.join(process.cwd(), config['schema-output'])

  // Create generation target folder, if it does not exist
  mkdirp.sync(path.dirname(outputPath))

  try {
    console.log(chalk.cyan('Generating new schema :', outputPath))
    fs.writeFileSync(outputPath, data, { flag: 'w' })
    return true
  } catch (error) {
    console.log(error)
    return error
  }
}
