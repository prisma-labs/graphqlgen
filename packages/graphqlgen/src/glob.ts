import * as glob from 'glob'
import { File } from 'graphqlgen-json-schema'
import { getPath } from './parse'

/**
 * Returns the path array from glob patterns
 */
export const extractGlobPattern = (file: File) => {
  return glob.sync(getPath(file))
}

/**
 * Handles the glob pattern of models.files
 */
export const handleGlobPattern = (files?: File[]): File[] => {
  if (!files) {
    return []
  }

  return files.reduce<File[]>((acc, file) => {
    const globedPaths = extractGlobPattern(file)

    if (globedPaths.length === 0) {
      return [...acc, file]
    }

    const globedFiles: File[] = globedPaths.map(path => {
      if (typeof file === 'string') {
        return path
      }

      return {
        path,
        defaultName: file.defaultName,
      }
    })

    return [...acc, ...globedFiles]
  }, [])
}
