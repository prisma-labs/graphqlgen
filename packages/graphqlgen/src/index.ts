#!/usr/bin/env node

import chalk from 'chalk'
import * as prettier from 'prettier'
import * as yargs from 'yargs'
import { GraphQLGenDefinition, Language } from 'graphqlgen-json-schema'
import { GraphQLTypes } from './source-helper'
import { IGenerator, GenerateArgs, CodeFileLike, ModelMap } from './types'
import {
  generate as generateTS,
  format as formatTS,
} from './generators/ts-generator'
import {
  generate as generateFlow,
  format as formatFlow,
} from './generators/flow-generator'
import { generate as scaffoldTS } from './generators/ts-scaffolder'
import { generate as scaffoldFlow } from './generators/flow-scaffolder'
import { parseConfig, parseContext, parseSchema, parseModels } from './parse'
import { validateConfig } from './validation'
import { handleGlobPattern } from './glob'
import * as Project from './project-output'

export type GenerateCodeArgs = {
  schema: GraphQLTypes
  config: GraphQLGenDefinition
  modelMap: ModelMap
  prettify?: boolean
  prettifyOptions?: prettier.Options
  language: Language
}

function getTypesGenerator(language: Language): IGenerator {
  switch (language) {
    case 'typescript':
      return { generate: generateTS, format: formatTS }
    case 'flow':
      return { generate: generateFlow, format: formatFlow }
  }
}

function getResolversGenerator(language: Language): IGenerator {
  switch (language) {
    case 'typescript':
      return { generate: scaffoldTS, format: formatTS }
    case 'flow':
      return { generate: scaffoldFlow, format: formatFlow }
  }
}

function generateTypes(
  generateArgs: GenerateArgs,
  generateCodeArgs: GenerateCodeArgs,
): string {
  const generatorFn: IGenerator = getTypesGenerator(generateCodeArgs.language!)
  const generatedTypes = generatorFn.generate(generateArgs)

  return generateCodeArgs.prettify
    ? generatorFn.format(
        generatedTypes as string,
        generateCodeArgs.prettifyOptions,
      )
    : (generatedTypes as string)
}

function generateResolvers(
  generateArgs: GenerateArgs,
  generateCodeArgs: GenerateCodeArgs,
): CodeFileLike[] {
  const generatorFn: IGenerator = getResolversGenerator(
    generateCodeArgs.language!,
  )
  const generatedResolvers = generatorFn.generate(
    generateArgs,
  ) as CodeFileLike[]

  return generatedResolvers.map(r => {
    return {
      path: r.path,
      force: r.force,
      code: generateCodeArgs.prettify
        ? generatorFn.format(r.code, generateCodeArgs.prettifyOptions)
        : r.code,
    }
  })
}

export function generateCode(
  generateCodeArgs: GenerateCodeArgs,
): { generatedTypes: string; generatedResolvers: CodeFileLike[] } {
  const { schema } = generateCodeArgs
  const generateArgs: GenerateArgs = {
    ...schema,
    context: parseContext(
      generateCodeArgs.config.context,
      generateCodeArgs.config.output,
    ),
    modelMap: generateCodeArgs.modelMap!,
  }
  const generatedTypes = generateTypes(generateArgs, generateCodeArgs)
  const generatedResolvers = generateResolvers(generateArgs, generateCodeArgs)
  // const generatedModels = generateModels(generateArgs, {schema, prettify, prettifyOptions, language})

  return { generatedTypes, generatedResolvers }
}

/**
 * The CLI interface
 */
async function run() {
  const argv = yargs
    .usage('Usage: graphqlgen or gg')
    .alias('i', 'init')
    .describe('i', 'Initialize a graphqlgen.yml file')
    .alias('v', 'version')
    .describe('v', 'Print the version of graphqlgen')
    .version()
    .strict()
    .help('h')
    .alias('h', 'help').argv

  if (argv.i) {
    Project.writeConfigScaffolding()
    return true
  }

  const config = parseConfig()
  const parsedSchema = parseSchema(config.schema)

  // Override the config.models.files using handleGlobPattern
  config.models = {
    ...config.models,
    files: handleGlobPattern(config.models.files),
  }

  if (!validateConfig(config, parsedSchema)) {
    return false
  }

  const modelMap = parseModels(
    config.models,
    parsedSchema,
    config.output,
    config.language,
  )

  const options = (await prettier.resolveConfig(process.cwd())) || {} // TODO: Abstract this TS specific behavior better

  if (JSON.stringify(options) !== '{}') {
    console.log(chalk.blue(`Found a prettier configuration to use`))
  }

  const { generatedTypes, generatedResolvers } = generateCode({
    schema: parsedSchema!,
    language: config.language,
    prettify: true,
    prettifyOptions: options,
    config,
    modelMap,
  })

  Project.writeTypes(generatedTypes, config)
  Project.writeResolversScaffolding(generatedResolvers, config)
}

// Only call run when running from CLI, not when included for tests
if (require.main === module) {
  run()
}
