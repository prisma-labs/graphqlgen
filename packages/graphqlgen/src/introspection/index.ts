import * as fs from 'fs'
import { Language } from 'graphqlgen-json-schema'

import { FilesToTypesMap, TypeDefinition, TypesMap } from './types'
import { buildTSTypesMap } from './ts-ast'
import { buildFlowTypesMap } from './flow-ast'
import { NormalizedFile } from '../parse'

export const filesToTypesMap: { [filePath: string]: TypesMap } = {}

function buildTypesMapByLanguage(
  fileContent: string,
  filePath: string,
  language: Language,
): TypesMap {
  switch (language) {
    case 'typescript':
      return buildTSTypesMap(fileContent, filePath)
    case 'flow':
      return buildFlowTypesMap(fileContent, filePath)
  }
}

export function addFileToTypesMap(
  filePath: string,
  language: Language,
): TypesMap {
  if (filesToTypesMap[filePath] !== undefined) {
    return filesToTypesMap[filePath]
  }

  const fileContent = fs.readFileSync(filePath).toString()

  const typesMap = buildTypesMapByLanguage(fileContent, filePath, language)

  filesToTypesMap[filePath] = typesMap

  return typesMap
}

export function findTypeInFile(
  filePath: string,
  typeName: string,
  language: Language,
): TypeDefinition | undefined {
  addFileToTypesMap(filePath, language)

  return filesToTypesMap[filePath][typeName]
}

export function addFilesToTypesMap(
  files: NormalizedFile[],
  language: Language,
): FilesToTypesMap {
  files.forEach(file => {
    addFileToTypesMap(file.path, language)
  })

  return filesToTypesMap
}

export function getFilesToTypesMap() {
  return filesToTypesMap
}
