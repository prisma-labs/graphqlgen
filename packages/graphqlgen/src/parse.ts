import * as Ajv from 'ajv'
import * as chalk from 'chalk'
import * as fs from 'fs'
import * as yaml from 'js-yaml'
import * as Path from 'path'
import * as tsNode from 'ts-node'
import { print } from 'graphql'
import { importSchema } from 'graphql-import'
import {
  GraphQLGenDefinition,
  Language,
  Models,
  File,
} from 'graphqlgen-json-schema'
import schema = require('graphqlgen-json-schema/dist/schema.json')

import { ContextDefinition, ModelMap, Model } from './types'
import {
  getAbsoluteFilePath,
  getImportPathRelativeToOutput,
} from './path-helpers'
import { getTypeToFileMapping, replaceAll, normalizeFilePath } from './utils'
import {
  extractTypes,
  extractGraphQLTypesWithoutRootsAndInputsAndEnums,
  GraphQLTypes,
} from './source-helper'
import { FilesToTypesMap } from './introspection/types'
import { getFilesToTypesMap } from './introspection'

export interface NormalizedFile {
  path: string
  defaultName?: string
}

const ajv = new Ajv().addMetaSchema(
  require('ajv/lib/refs/json-schema-draft-06.json'),
)
const validateYaml = ajv.compile(schema)

export function parseConfig(): GraphQLGenDefinition {
  if (!fs.existsSync('graphqlgen.yml')) {
    console.error(chalk.default.red(`No graphqlgen.yml found`))
    process.exit(1)
  }

  const config = yaml.safeLoad(
    fs.readFileSync('graphqlgen.yml', 'utf-8'),
  ) as GraphQLGenDefinition

  if (!validateYaml(config)) {
    console.error(chalk.default.red(`Invalid graphqlgen.yml file`))
    console.error(chalk.default.red(printErrors(validateYaml.errors!)))
    process.exit(1)
  }

  return config
}

function printErrors(errors: any[]): string {
  return errors
    .map(e => {
      const params = Object.keys(e.params)
        .map(key => `${key}: ${e.params[key]}`)
        .join(', ')
      return `${e.dataPath} ${e.message}. ${params}`
    })
    .join('\n')
}

export function parseContext(
  context: string | undefined,
  outputDir: string,
): ContextDefinition | undefined {
  if (!context) {
    return undefined
  }

  const [filePath, interfaceName] = context.split(':')

  return {
    contextPath: getImportPathRelativeToOutput(filePath, outputDir),
    interfaceName,
  }
}

export function parseSchema(schemaPath: string): GraphQLTypes {
  const [filePath, exportName = 'default'] = schemaPath.split(':')

  // We can assume absolute path is cwd prefixed because
  // gg currently only works when run in a directory with the
  // graphqlgen manifest.
  const absoluteFilePath =
    filePath[0] == '/' ? filePath : Path.join(process.cwd(), filePath)

  if (!fs.existsSync(absoluteFilePath)) {
    console.error(
      chalk.default.red(`The schema file ${filePath} does not exist`),
    )
    process.exit(1)
  }

  let schema: string | undefined
  try {
    if (filePath.endsWith('.ts')) {
      tsNode.register({
        transpileOnly: true,
      })
      const schemaModule = require(absoluteFilePath)
      const loadedSchema = schemaModule[exportName]

      if (typeof loadedSchema === 'string') {
        schema = loadedSchema
      } else {
        schema = print(loadedSchema)
      }
    } else {
      schema = importSchema(filePath)
    }
  } catch (e) {
    console.error(
      chalk.default.red(`Error occurred while reading schema: ${e}`),
    )
    process.exit(1)
  }

  let types: GraphQLTypes
  try {
    types = extractTypes(schema!)
  } catch (e) {
    console.error(chalk.default.red(`Failed to parse schema: ${e}`))
    process.exit(1)
  }

  return types!
}

function buildModel(
  modelName: string,
  filePath: string,
  filesToTypesMap: FilesToTypesMap,
  outputDir: string,
  language: Language,
): Model {
  const absoluteFilePath = getAbsoluteFilePath(filePath, language)
  const importPathRelativeToOutput = getImportPathRelativeToOutput(
    absoluteFilePath,
    outputDir,
  )
  return {
    absoluteFilePath,
    importPathRelativeToOutput,
    definition: filesToTypesMap[filePath][modelName],
  }
}

export function getPath(file: File): string {
  if (typeof file === 'string') {
    return file
  }

  return file.path
}

export function getDefaultName(file: File): string | null {
  if (typeof file === 'string') {
    return null
  }

  return file.defaultName || null
}

export function normalizeFiles(
  files: File[] | undefined,
  language: Language,
): NormalizedFile[] {
  return files !== undefined
    ? files.map(file => ({
        defaultName: typeof file === 'object' ? file.defaultName : undefined,
        path: normalizeFilePath(getPath(file), language),
      }))
    : []
}

export function parseModels(
  models: Models,
  schema: GraphQLTypes,
  outputDir: string,
  language: Language,
): ModelMap {
  const graphQLTypes = extractGraphQLTypesWithoutRootsAndInputsAndEnums(schema)
  const normalizedFiles = normalizeFiles(models.files, language)
  const filesToTypesMap = getFilesToTypesMap()
  const overriddenModels = !!models.override ? models.override : {}
  const typeToFileMapping = getTypeToFileMapping(
    normalizedFiles,
    filesToTypesMap,
  )

  return graphQLTypes.reduce((acc, type) => {
    if (overriddenModels[type.name]) {
      const [filePath, modelName] = models.override![type.name].split(':')

      return {
        ...acc,
        [type.name]: buildModel(
          modelName,
          normalizeFilePath(filePath, language),
          filesToTypesMap,
          outputDir,
          language,
        ),
      }
    }

    const typeFileTuple = Object.entries(typeToFileMapping).find(
      ([typeName, file]) => {
        const defaultName = getDefaultName(file)
        const replacedTypeName = defaultName
          ? replaceVariablesInString(defaultName, { typeName: type.name })
          : type.name

        return typeName === replacedTypeName
      },
    )

    if (!typeFileTuple) {
      throw new Error(
        `Could not find type ${type.name} in any of the provided files`,
      )
    }

    const file = typeFileTuple[1]

    const filePath = getPath(file)
    const defaultName = getDefaultName(file)

    const replacedTypeName = defaultName
      ? replaceVariablesInString(defaultName, { typeName: type.name })
      : type.name

    return {
      ...acc,
      [type.name]: buildModel(
        replacedTypeName,
        filePath,
        filesToTypesMap,
        outputDir,
        language,
      ),
    }
  }, {})
}

interface ReplacementMap {
  [key: string]: string
}

export function replaceVariablesInString(
  str: string,
  replacements: ReplacementMap,
) {
  const variableSyntax = RegExp(
    '\\${([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)]+?)}',
    'g',
  )
  let newStr = str
  if (variableSyntax.test(str)) {
    str.match(variableSyntax)!.forEach(matchedString => {
      // strip ${} away to get the pure variable name
      const variableName = matchedString
        .replace(variableSyntax, (_, varName) => varName.trim())
        .replace(/\s/g, '')

      if (replacements[variableName]) {
        newStr = replaceAll(newStr, matchedString, replacements[variableName])
      } else {
        throw new Error(`Variable ${variableName} is not covered by a value`)
      }
    })
  }

  return newStr
}
