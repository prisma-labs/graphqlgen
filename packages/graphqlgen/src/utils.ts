import * as Path from 'path'
import * as FS from 'fs'
import { Language } from 'graphqlgen-json-schema'

import { getExtNameFromLanguage } from './path-helpers'
import { NormalizedFile } from './parse'
import { FilesToTypesMap, InterfaceNamesToFile } from './introspection/types'

/**
 * Uppercase the first letter of a string. Useful when generating type names.
 */
const upperFirst = (s: string): string => {
  return s.replace(/^\w/, c => c.toUpperCase())
}

/**
 * Append a file extension to a file name. Leading dots in given
 * file extension are gracefully dropped.
 */
const appendExt = (ext: string, filePath: string): string => {
  const normalizedExt = ext.replace(/^\.+/, '')
  return filePath + '.' + normalizedExt
}

/**
 * Normalize different kinds of path notation. Will do synchronous
 * file IO to accomplish task.
 *
 * Examples:
 *
 *    ./path/to/index.ts => `pwd`/path/to/index.ts
 *    ./path/to          => `pwd`/path/to/to.ts
 *    ./path/to          => `pwd`/path/to/index.ts
 *    ./path/to/         => `pwd`/path/to/index.ts
 */
const normalizeFilePath = (filePath: string, language: Language): string => {
  const ext = getExtNameFromLanguage(language)

  // If the filepath is set against a file then just return that.
  if (Path.extname(filePath) === ext) {
    return Path.resolve(filePath)
  }

  // If there is no file extension then infer it (from given language) and return if
  // exists on file system. Otherwise consider file path as being to a folder that
  // mnust have an index file.

  const filePathWithExt = Path.resolve(filePath) + ext

  if (FS.existsSync(filePathWithExt)) {
    return filePathWithExt
  }

  return Path.join(Path.resolve(filePath), appendExt(ext, 'index'))
}

/**
 * Create a map of interface names to the path of the file in which they're defined
 * The first evaluated interfaces are always the chosen ones
 */
const getTypeToFileMapping = (
  files: NormalizedFile[],
  filesToTypesMap: FilesToTypesMap,
): InterfaceNamesToFile => {
  // REFACTOR: This function basically just takes an index and flips it. Make generic.
  const mapping: InterfaceNamesToFile = {}

  for (const file of files) {
    // WARNING: typesMap is not typesafe since the lookup could fail.
    const typesMap = filesToTypesMap[file.path]
    const interfaceNames = Object.keys(typesMap)

    for (const interfaceName of interfaceNames) {
      if (!mapping[interfaceName]) {
        mapping[interfaceName] = file
      }
    }
  }

  return mapping
}

/**
 * Replace all occurances of given search string in a given
 * string with another string.
 */
const replaceAll = (
  str: string,
  search: string,
  replacement: string,
): string => {
  return str.split(search).join(replacement)
}

/**
 * Return a new array whose items are a merger of the given two arrays.
 */
const concat = <T = unknown>(a: T[], b: T[]): T[] => {
  return a.concat(b)
}

// TODO Refactor; confusing; only one callsite
const uniq = <T = unknown>(value: T, index: number, array: T[]): boolean => {
  return array.indexOf(value) === index
}

type LanguageExtension = 'ts' | 'js'
const languageExtensions: LanguageExtension[] = ['ts', 'js']

const extToLangIndex: Record<'ts' | 'js', Language> = {
  ts: 'typescript',
  js: 'flow',
}

const langToExtIndex: Record<Language, 'ts' | 'js'> = {
  typescript: 'ts',
  flow: 'js',
}

const getExtFromLang = (lang: Language): LanguageExtension => {
  return langToExtIndex[lang]
}

const getLangFromExt = (ext: LanguageExtension): Language => {
  return extToLangIndex[ext]
}

/**
 * Get the extension from a file name (or path with file name).
 * Unlike Path.extname this returns null if no ext can be extracted.
 */
const getExt = (path: string): null | string => {
  const ext = Path.extname(path)
  if (ext === '') return null
  const extWithoutDot = ext.slice(1)
  return extWithoutDot
}

/**
 * Predicate function for checking if a path is to file.
 * Relies on convention that a dot being present in last
 * path item is a file.
 *
 * Examples:
 *
 *    /a/b/c -> false
 *    /a/b/c.foo -> true
 */
const isFile = (path: string): boolean => {
  return Path.extname(path) !== ''
}

export {
  LanguageExtension,
  languageExtensions,
  getLangFromExt,
  getExtFromLang,
  getTypeToFileMapping,
  uniq,
  concat,
  replaceAll,
  upperFirst,
  normalizeFilePath,
  getExt,
  isFile,
}
