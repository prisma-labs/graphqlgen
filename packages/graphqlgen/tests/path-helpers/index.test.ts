import { getAbsoluteFilePath } from '../../src/path-helpers'
import { join, resolve } from 'path'
import { normalizeFilePath } from '../../src/utils'

const relative = (p: string) => join(__dirname, p)
const language = 'typescript'

test('invalid path', () => {
  expect(() =>
    getAbsoluteFilePath(join('../fixtures/type.ts'), language),
  ).toThrow()
})

test('valid path', () => {
  expect(
    getAbsoluteFilePath(join(__dirname, '../fixtures/basic/index.ts'), language)
      .split('/')
      .slice(-4),
  ).toMatchSnapshot()
})

test('valid directory path', () => {
  expect(
    getAbsoluteFilePath(relative('../fixtures/basic/'), language)
      .split('/')
      .slice(-4),
  ).toMatchSnapshot()
})

test('valid directory path without slash', () => {
  expect(
    getAbsoluteFilePath(relative('../fixtures/basic'), language)
      .split('/')
      .slice(-4),
  ).toMatchSnapshot()
})

test('normalizeFilePath()', () => {
  // '/path/to/file.ts' => '/path/to/file.ts'
  expect(normalizeFilePath(relative('./mocks/types.ts'), language)).toEqual(
    resolve(relative('./mocks/types.ts')),
  )

  // '/path/to/file' => '/path/to/file.ts'
  expect(normalizeFilePath(relative('./mocks/types'), language)).toEqual(
    resolve(relative('./mocks/types.ts')),
  )

  // '/path/to/file' => '/path/to/file/index.ts'
  expect(normalizeFilePath(relative('./mocks/dir-1'), language)).toEqual(
    resolve(relative('./mocks/dir-1/index.ts')),
  )

  // '/path/to/file' => '/path/to/file.ts'
  // Even though there is a folder of the same name
  expect(normalizeFilePath(relative('./mocks/dir-2/dir-3'), language)).toEqual(
    resolve(relative('./mocks/dir-2/dir-3.ts')),
  )
})
