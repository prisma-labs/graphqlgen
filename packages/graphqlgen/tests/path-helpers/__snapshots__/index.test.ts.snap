// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valid directory path 1`] = `
Array [
  "tests",
  "fixtures",
  "basic",
  "index.ts",
]
`;

exports[`valid directory path without slash 1`] = `
Array [
  "tests",
  "fixtures",
  "basic",
  "index.ts",
]
`;

exports[`valid path 1`] = `
Array [
  "tests",
  "fixtures",
  "basic",
  "index.ts",
]
`;
